#STEP 2
#Generate the individual points for geographical cline fitting based on individual SLIM replicates
#input: SLIM replicates output generated by slim_replicates.R
#output: input table for hzar

library(dplyr)
library(tidyverse)
library(foreach)
library(doParallel)
#setwd("C:/Users/230982/Dropbox/MHC_projekt/Shared_MHC_project/General_analysis/SLIM_simulation")
#setwd("~/Dropbox/MHC_projekt/Shared_MHC_project/General_analysis/SLIM_simulation/reduced_mutation_rate")
#
### FUNCTIONS ####
get_mutations <- function(slim_output){
  input <- readLines(slim_output)
  mutation_index <- which(grepl("Mutations:",input))
  ending_index <- which(grepl(":$",input)) %>% .[which(.==mutation_index)+1]-1
  
  dt <- input[(mutation_index+1):ending_index] %>%
    paste(collapse = "\n") %>% read.delim(text = .,sep = " ",header = F) %>% select(-V2)
  colnames(dt) <- c("MUTATION_ID","MUTATION_TYPE","POSITION","s","h","SUBPOPULATION","TICK_ORIGIN","N_COPIES")
  dt
}

get_individuals <- function(slim_output){
  input <- readLines(slim_output)
  mutation_index <- which(grepl("Individuals:",input))
  ending_index <- which(grepl(":$",input)) %>% .[which(.==mutation_index)+1]-1
  
  dt <- input[(mutation_index+1):ending_index] %>%
    paste(collapse = "\n") %>% read.delim(text = .,sep = c(" "),header = F) %>%
    separate(1,into = c("SUBPOPULATION","INDIVIDUAL_ID"),sep = ":") %>%
    #mutate(across(.cols= c(V3,V4),.fns = function(x){stringr::str_replace_all(string = x,pattern = "p[1,2]:",replacement = "")})) %>%
    pivot_longer(cols = c("V3","V4"),names_to = "COL",values_to = "GENOMES") %>%
    select(-COL) %>% rename("SEX" = "V2")
  dt
}

get_genomes <- function(slim_output){
  input <- readLines(slim_output)
  mutation_index <- which(grepl("Genomes:",input))
  
  dt <- input[(mutation_index+1):length(input)]
  
  col_n <- str_split(dt," ") %>% sapply(length) %>% max
  
  dt <- dt %>% paste(collapse = "\n") %>% 
    read.table(text = .,sep = c(" "),header = F,fill = T,col.names = paste0("V",1:col_n)) %>%
    select(-V2) %>% rename("GENOMES" = "V1") %>%
    pivot_longer(cols = -1,names_to = NULL,values_to = "MUTATION_ID",values_drop_na = T) 
    
  dt
}

h_max_likelihood_standard <- function(h,dt){ 
  vec <- c()
  for(i in unique(dt$POSITION)){
    temp2 <- filter(dt,POSITION == i)
    if(length(unique(temp2$MUTATION_ID))>1){ #if heterozygote
      p1 <- (h*temp2$p2[1])+((1-h)*temp2$p1[1])
      p2 <- (h*temp2$p2[2])+((1-h)*temp2$p1[2])
      vec <- c(vec,log(2*p1*p2))
    } else { #if homozygote
      p <- (h*temp2$p2[1])+((1-h)*temp2$p1[1])
      vec <- c(vec,log((p^2)))
    }
  }
  vec <- vec[!is.infinite(vec)] #<!> solution
  if(length(vec) == 0){
    return(NA)
  } else {
    #print(sum(vec))
    sum(vec)
  }
} #calculates h-index, it must be run with optimization

run_h <- function(x){
  tryCatch(optimize(h_max_likelihood_standard,dt = x,interval = c(0,1),maximum = T)$maximum,
                warning = function(w){return(NA)})
}

run_h_PA <- function(temp){
  tryCatch(optimize(h_max_likelihood_PA,present = temp$GENOTYPE,s = temp$p1,r = temp$p2,interval = c(0,1),maximum = T)$maximum,
           warning = function(w){return(NA)})
}

h_max_likelihood_PA <- function(h,present,s,r){ 
  #H-W frequency of allele (observed values reflect phenotype)
  s_null <- sqrt(1-s)
  r_null <- sqrt(1-r)
  s_full <- 1-(sqrt(1-s))
  r_full <- 1-(sqrt(1-r))
  
  #probabilities
  p1 <- (r_full*h)+((1-h)*s_full)
  p2 <- (r_null*h)+((1-h)*s_null)
  
  # <!> There is a problem here <!>
  # -Inf is returned if a allele is absent in both species or has frequency 1 in both species
  vec <- c()
  for(i in 1:length(present)){
    if(present[i] == 1){
      vec <- c(vec,log(p1[i]^2 + (2*(p1[i]*p2[i]))))
    } else {vec <- c(vec,log(p2[i]^2))}
  }
  vec <- vec[!is.infinite(vec)] #<!> solution
  if(length(vec) == 0){
    return(NA)
  } else {
    #print(sum(vec))
    sum(vec)
  }
} #calculates h-index for PA data, it must be run with optimization

sigmoidal <- function(x,c,w){
  1/(1+exp((-4*(x-c))/w))
}

dir_list <- list.dirs(recursive = F) %>% .[. != "."]

for(d in dir_list){

#d <- "selection_NFS_theta_2x_replicate_1"
  
dir.create(paste0(d,"/tables"))
dir.create(paste0(d,"/plots"))

#dt - combined SLIM
#dt_s - sample for frequency estimation
#al_frq - allele/phenotype frequencies based on whole population
#al_frq_sample - allele/phenotype frequencies based on the sample
#offspring - generated offspring (later subsampled)
#pedigrees - H_index based on pedigrees

### DATA PREPARATION ####
out_list <- list()
for(i in 0:4){
  slim_output <- paste0(d,"/original/output_",i,".txt")
  muts <- get_mutations(slim_output) %>% select(MUTATION_ID,POSITION,N_COPIES)
  inds <- get_individuals(slim_output) %>% select(-SEX)
  genomes <- get_genomes(slim_output)
  
  out_list[[length(out_list)+1]] <- left_join(inds,genomes,by = "GENOMES") %>% 
    left_join(muts,by = "MUTATION_ID") %>% 
    mutate(STAGE = paste0("stage_",i)) %>%
    group_by(SUBPOPULATION,INDIVIDUAL_ID,STAGE) %>%
    mutate("N" = n()) %>% #remove all individuals who does not have full genomes
    filter(N == 10) %>% select(-N)
}
dt <- as.data.frame(do.call(rbind,out_list))
write.csv(dt,paste0(d,"/tables/SLIM_combined.csv"),row.names = F,fileEncoding = "UTF-8")
rm("genomes","inds","muts","out_list")
#
### ANALYSIS ####
dt <- arrange(dt,STAGE,POSITION,SUBPOPULATION)

#taking samples
get_sample <- function(x,size){
  to_sample <- sample(unique(x$INDIVIDUAL_ID),size = size,replace = F)
  filter(x,INDIVIDUAL_ID %in% to_sample)
}

dt_s <- dt %>% group_by(STAGE,SUBPOPULATION) %>% get_sample(size = 30) %>% ungroup()
write.csv(dt_s,paste0(d,"/tables/sample_for_frequencies.csv"),row.names = F,fileEncoding = "UTF-8")

#descriptive
dt_s <- arrange(dt_s,STAGE,POSITION,SUBPOPULATION)
temp <- dt_s %>% group_by(STAGE,SUBPOPULATION) %>%
  mutate("N_ALLELES" = length(unique(MUTATION_ID))) %>% #group_by(STAGE) %>%
  mutate("N_SHARED" = sum(unique(.$MUTATION_ID[.$SUBPOPULATION == unique(SUBPOPULATION) & .$STAGE == unique(STAGE)]) %in% .$MUTATION_ID[.$SUBPOPULATION != unique(SUBPOPULATION) & .$STAGE == unique(STAGE)])) %>%
  mutate("PERC_SHARED" = round(N_SHARED/N_ALLELES*100,2)) %>%
  select(SUBPOPULATION,STAGE,N_ALLELES,N_SHARED,PERC_SHARED) %>% distinct()

write.csv(temp,paste0(d,"/tables/descriptive.csv"),row.names = F,fileEncoding = "UTF-8")

#descriptive per locus
temp <- dt_s %>% group_by(STAGE,SUBPOPULATION,POSITION) %>%
  mutate("N_ALLELES" = length(unique(MUTATION_ID))) %>% #group_by(STAGE) %>%
  mutate("N_SHARED" = sum(unique(.$MUTATION_ID[.$SUBPOPULATION == unique(SUBPOPULATION) & .$STAGE == unique(STAGE) & .$POSITION == unique(POSITION)]) %in%
                            .$MUTATION_ID[.$SUBPOPULATION != unique(SUBPOPULATION) & .$STAGE == unique(STAGE) & .$POSITION == unique(POSITION)])) %>%
  mutate("PERC_SHARED" = round(N_SHARED/N_ALLELES*100,2)) %>%
  select(SUBPOPULATION,STAGE,POSITION,N_ALLELES,N_SHARED,PERC_SHARED) %>% distinct() %>%
  arrange(STAGE,POSITION,SUBPOPULATION)
write.csv(temp,paste0(d,"/tables/descriptive_locus.csv"),row.names = F,fileEncoding = "UTF-8")

temp <- temp %>% pivot_longer(cols = starts_with("N_"),names_to = "TYPE",values_to = "N_ALLELES")

p <- ggplot(temp,aes(x = POSITION, y = N_ALLELES,fill = TYPE))+
  geom_col(col = "transparent",alpha = 0.7,position = position_identity())+
  theme_bw()+
  facet_grid(SUBPOPULATION ~ STAGE)
ggsave(paste0(d,"/plots/N_alleles_perLocus.png"),device = "png",plot = p)

#allele frequencies - whole population
al_frq <- dt %>% group_by(SUBPOPULATION,STAGE,POSITION) %>%
  mutate("N_BP" = n()) %>% group_by(SUBPOPULATION,STAGE,POSITION,MUTATION_ID) %>%
  mutate("N_COPIES" = n()) %>% ungroup() %>%
  mutate("FREQ" = N_COPIES/N_BP) %>%
  select(SUBPOPULATION,STAGE,MUTATION_ID,FREQ) %>% distinct() %>%
  complete(SUBPOPULATION,STAGE,MUTATION_ID,fill = list("FREQ" = 0))
  
write.csv(al_frq,paste0(d,"/tables/frequency_wholePop.csv"),row.names = F,fileEncoding = "UTF-8")

#phenotype frequencies
al_frq_ph <- dt %>% group_by(SUBPOPULATION,STAGE,INDIVIDUAL_ID,MUTATION_ID) %>%
  sample_n(1) %>% group_by(SUBPOPULATION,STAGE) %>%
  mutate("N_BP" = n()) %>% group_by(SUBPOPULATION,STAGE,MUTATION_ID) %>%
  mutate("N_COPIES" = n()) %>% ungroup() %>%
  mutate("FREQ" = N_COPIES/N_BP) %>%
  select(SUBPOPULATION,STAGE,MUTATION_ID,FREQ) %>% distinct() %>%
  complete(SUBPOPULATION,STAGE,MUTATION_ID,fill = list("FREQ" = 0))

write.csv(al_frq_ph,paste0(d,"/tables/frequency_wholePop_Phenotype.csv"),row.names = F,fileEncoding = "UTF-8")

#allele frequencies - samples
al_frq_sample <- dt_s %>% group_by(SUBPOPULATION,STAGE,POSITION) %>%
  mutate("N_BP" = n()) %>% group_by(SUBPOPULATION,STAGE,POSITION,MUTATION_ID) %>%
  mutate("N_COPIES" = n()) %>% ungroup() %>%
  mutate("FREQ" = N_COPIES/N_BP) %>%
  select(SUBPOPULATION,STAGE,MUTATION_ID,FREQ) %>% distinct() %>%
  complete(SUBPOPULATION,STAGE,MUTATION_ID,fill = list("FREQ" = 0))

write.csv(al_frq_sample,paste0(d,"/tables/frequency_sample.csv"),row.names = F,fileEncoding = "UTF-8")

#phenotype frequencies - samples
al_frq_sample_ph <- dt_s %>% group_by(SUBPOPULATION,STAGE,INDIVIDUAL_ID,MUTATION_ID) %>%
  sample_n(1) %>% group_by(SUBPOPULATION,STAGE) %>%
  mutate("N_BP" = n()) %>% group_by(SUBPOPULATION,STAGE,MUTATION_ID) %>%
  mutate("N_COPIES" = n()) %>% ungroup() %>%
  mutate("FREQ" = N_COPIES/N_BP) %>%
  select(SUBPOPULATION,STAGE,MUTATION_ID,FREQ) %>% distinct() %>%
  complete(SUBPOPULATION,STAGE,MUTATION_ID,fill = list("FREQ" = 0))

write.csv(al_frq_sample_ph,paste0(d,"/tables/frequency_sample_Phenotype.csv"),row.names = F,fileEncoding = "UTF-8")

# #frequencies correlation
# temp <- al_frq %>% rename("FREQ_WHOLE" = "FREQ") %>% 
#   inner_join(al_frq_sample,by = c("SUBPOPULATION","STAGE","MUTATION_ID"))
# 
# #ggplot(filter(fr_cor,SUBPOPULATION == "p1"),aes(x = FREQ_WHOLE,y = FREQ))+
# p <- ggplot(temp,aes(x = FREQ_WHOLE,y = FREQ))+
#   geom_point()+
#   theme_bw()+geom_smooth(method = "lm")+labs(x = "Frequencies in whole population", y = "Frequencies in sample")+
#   facet_grid(cols = vars(STAGE))
# ggsave(paste0(d,"/plots/Freqs_correlations.png"),device = "png",plot = p)
#
### GENERATE NEDEED ADMIXTURE VALUES FROM SIGMOIDAL SHAPE ####
my_dist <- sort(c(sample(1:97,size = 10,replace = F), #left tail
             sample(seq(97,103,0.1),size = 15,replace = F), #centre
             sample(103:200,size = 10,replace = F)))  #linear tail

#cline parameters
center <- 100
width <- 6

chosen_contributions <- round(sigmoidal(my_dist,center,width),3)
dist_tab <- data.frame("DISTANCE" = my_dist,"POP_ID" = paste0("pop",1:length(my_dist)))
#
### OFFSPRING CREATION ####
n_offspring <- 10
out_list <- list()

for(s in unique(dt$STAGE)){
  gen_p1 <- filter(dt,STAGE == s & SUBPOPULATION == "p1") %>% pull(GENOMES) %>% unique() 
  gen_p2 <- filter(dt,STAGE == s & SUBPOPULATION == "p2") %>% pull(GENOMES) %>% unique() 
  
  for(i in 1:length(chosen_contributions)){ #step of cline
    for(j in 1:n_offspring){ #each kid
      
      ind_genotype <- sample(c("p1","p2"),size = 2,replace = T,prob = c(1-chosen_contributions[i],chosen_contributions[i]))
      chosen_genomes <- c()
      for(g in 1:length(ind_genotype)){
        chosen_genomes[g] <- ifelse(ind_genotype[g] == "p1",yes = sample(gen_p1,1),no = sample(gen_p2,1)) #both genomes of an offspring
      }
      out_list[[length(out_list)+1]] <- c(s,chosen_contributions[i],paste0("pop",i),paste0("o",j),chosen_genomes)
    }
  }
}
offspring <- as.data.frame(do.call(rbind,out_list))
colnames(offspring) <- c("STAGE","p2_CONTRIBUTION","POP_ID","OFFSPRING_ID","GENOME1","GENOME2")
offspring <- pivot_longer(offspring,cols = c("GENOME1","GENOME2"),names_to = "GENOME_ORDER",values_to = "GENOMES") %>%
  left_join((select(dt,GENOMES,MUTATION_ID,STAGE,POSITION)),by = c("STAGE","GENOMES")) %>%
  mutate("ORIGIN" = ifelse(grepl(x = GENOMES,pattern = "p1:"),"p1","p2"))
  
write.csv(offspring,paste0(d,"/tables/offspring.csv"),row.names = F,fileEncoding = "UTF-8")
rm("out_list")
#
### H INDEX  - PEDIGREES ####
#known pedigrees
pedigrees <- offspring %>% select(STAGE,p2_CONTRIBUTION,POP_ID,OFFSPRING_ID,POSITION,ORIGIN) %>%
  group_by(STAGE,p2_CONTRIBUTION,POP_ID,OFFSPRING_ID,POSITION) %>% mutate("LOCUS" = c("L1","L2")) %>%
  pivot_wider(values_from = "ORIGIN",names_from = LOCUS) %>%
  left_join(y = data.frame("L1" = c("p1","p1","p2","p2"),"L2" = c("p1","p2","p1","p2"),"H_INDEX" = c(0,.5,.5,1)),by = c("L1","L2")) %>% #much faster
  group_by(STAGE,p2_CONTRIBUTION,POP_ID,OFFSPRING_ID) %>%
  summarise("H_INDEX" = mean(H_INDEX)) %>%
  group_by(STAGE,p2_CONTRIBUTION,POP_ID) %>% summarise("H_INDEX" = mean(H_INDEX))

pedigrees <- rbind(mutate(pedigrees,"FREQ_TYPE" = "FREQ_SAMPLE"),mutate(pedigrees,"FREQ_TYPE" = "FREQ_WHOLE"))
  
write.csv(pedigrees,paste0(d,"/tables/H_index_knownPedigrees.csv"),row.names = F,fileEncoding = "UTF-8")
#
### H INDEX - FULL GENOTYPES ####
al_frq  <- al_frq %>% rename("FREQ_WHOLE" = "FREQ")
al_frq_sample <- al_frq_sample %>% rename("FREQ_SAMPLE" = "FREQ")

dt_temp <- left_join(offspring,al_frq,by = c("STAGE","MUTATION_ID")) %>%
  left_join(al_frq_sample,by = c("STAGE","MUTATION_ID","SUBPOPULATION")) %>%
  replace_na(list("FREQ_WHOLE" = 0,"FREQ_SAMPLE" = 0)) %>% 
  pivot_longer(cols = starts_with("FREQ"),names_to = "FREQ_TYPE",values_to = "FREQ") %>%
  pivot_wider(names_from = SUBPOPULATION,values_from = "FREQ") %>%
  mutate("IDENTIFIER" = paste(FREQ_TYPE,STAGE,p2_CONTRIBUTION,POP_ID,OFFSPRING_ID,sep = "&")) %>%
  filter(FREQ_TYPE == "FREQ_SAMPLE") #limit to sampled

## PARALLEL ###
writeLines("",paste0(d,"/log.txt"))

no_cores <- 14 #number of cores to use
cl <- makeCluster(no_cores)
registerDoParallel(cl)

genotypes <- foreach(i = unique(dt_temp$IDENTIFIER),.packages = "dplyr",.combine = "rbind") %dopar% {
    sink(paste0(d,"/log.txt"),append = T)
    print(i)
    sink()
    temp <- filter(dt_temp,IDENTIFIER == i)
    c(strsplit(unique(temp$IDENTIFIER),"&")[[1]],run_h(temp))
  }
stopCluster(cl)

rownames(genotypes) <- NULL
colnames(genotypes) <- c("FREQ_TYPE","STAGE","p2_CONTRIBUTION","POP_ID","OFFSPRING_ID","H_INDEX")
genotypes <- as.data.frame(genotypes) %>% na.omit() %>% mutate("H_INDEX" = as.numeric(H_INDEX))
write.csv(genotypes,paste0(d,"/tables/H_index_fullGenotypes.csv"),row.names = F,fileEncoding = "UTF-8")
#
### ENCODED GENOTYPES ####
al_frq_ph  <- al_frq_ph %>% rename("FREQ_WHOLE" = "FREQ")
al_frq_sample_ph <- al_frq_sample_ph %>% rename("FREQ_SAMPLE" = "FREQ")

dt_temp <- offspring %>% select(-POSITION,-ORIGIN,-GENOMES,-GENOME_ORDER) %>% distinct() %>%
  mutate("GENOTYPE" = 1) %>%
  pivot_wider(names_from = MUTATION_ID,values_from = GENOTYPE,values_fill = 0) %>%
  pivot_longer(cols = -c(1:4),names_to = "MUTATION_ID",values_to = "GENOTYPE") %>% arrange(STAGE,POP_ID,p2_CONTRIBUTION,MUTATION_ID) %>%
  mutate("MUTATION_ID" = as.integer(MUTATION_ID)) %>%
  left_join(al_frq_ph,by = c("STAGE","MUTATION_ID")) %>%
  left_join(al_frq_sample_ph,by = c("STAGE","MUTATION_ID","SUBPOPULATION")) %>%
  pivot_longer(cols = starts_with("FREQ"),names_to = "FREQ_TYPE",values_to = "FREQ") %>%
  pivot_wider(names_from = SUBPOPULATION,values_from = "FREQ") %>%
  replace_na(list("p1" = 0,"p2" = 0)) %>%
  mutate("IDENTIFIER" = paste(FREQ_TYPE,STAGE,p2_CONTRIBUTION,POP_ID,OFFSPRING_ID,sep = "&")) %>%
  filter(FREQ_TYPE == "FREQ_SAMPLE") #limit to sampled

## PARALLEL ###
writeLines("",paste0(d,"/log.txt"))

no_cores <- 14 #number of cores to use
cl <- makeCluster(no_cores)
registerDoParallel(cl)

encoded <- foreach(i = unique(dt_temp$IDENTIFIER),.packages = "dplyr",.combine = "rbind") %dopar% {
  sink(paste0(d,"/log.txt"),append = T)
  print(i)
  sink()
  temp <- filter(dt_temp,IDENTIFIER == i)
  
  #c(strsplit(unique(temp$IDENTIFIER),"&")[[1]],tryCatch(optimize(h_max_likelihood_PA,present = temp$GENOTYPE,s = temp$p1,r = temp$p2,interval = c(0,1),maximum = T)$maximum),warning = function(w){return(NA)})
  c(strsplit(unique(temp$IDENTIFIER),"&")[[1]],run_h_PA(temp))
  }
stopCluster(cl)

rownames(encoded) <- NULL
colnames(encoded) <- c("FREQ_TYPE","STAGE","p2_CONTRIBUTION","POP_ID","OFFSPRING_ID","H_INDEX")
encoded <- as.data.frame(encoded) %>% na.omit() %>% mutate("H_INDEX" = as.numeric(H_INDEX))
write.csv(encoded,paste0(d,"/tables/H_index_Encoded.csv"),row.names = F,fileEncoding = "UTF-8")
#
### TABLE FOR H-INDEX ####
#combine all h-index estimates
dt_genotypes <- genotypes %>%
  group_by(FREQ_TYPE,STAGE,p2_CONTRIBUTION,POP_ID) %>% summarise("H_INDEX_GENO" = mean(H_INDEX))
dt_encoded <- encoded %>%
  group_by(FREQ_TYPE,STAGE,p2_CONTRIBUTION,POP_ID) %>% summarise("H_INDEX_ENCO" = mean(H_INDEX))

dt_temp <- dt_genotypes %>%
  full_join(pedigrees,by = c("STAGE","p2_CONTRIBUTION","FREQ_TYPE","POP_ID")) %>%
  filter(FREQ_TYPE %in% unique(dt_genotypes$FREQ_TYPE)) %>%
  full_join(dt_encoded,by = c("STAGE","p2_CONTRIBUTION","FREQ_TYPE","POP_ID"))

h_index <- dt_temp %>% left_join(dist_tab,by = "POP_ID") %>%
  mutate(across(starts_with("H_"),.fns = function(x){x*100})) %>%
  arrange(FREQ_TYPE,STAGE,POP_ID,p2_CONTRIBUTION) %>%
  group_by(FREQ_TYPE,STAGE) %>%
  mutate("H_INDEX_N" = 20,"H_INDEX_GENO_N" = 20,
         "H_INDEX_ENCO_N" = 20,"HYBRID_ZONE" = paste(FREQ_TYPE,STAGE,sep = "_")) %>% ungroup() %>%
  select(HYBRID_ZONE,POP_ID,DISTANCE,H_INDEX,H_INDEX_N,H_INDEX_GENO,H_INDEX_GENO_N,H_INDEX_ENCO,H_INDEX_ENCO_N)

write.csv(h_index,paste0(d,"/tables/hzar_input.csv"),fileEncoding = "UTF-8",row.names = F)
}
